File: rag/scripts/ingest.py
"""
ingest.py — 知识文件扫描预览

用法：
    python rag/scripts/ingest.py
    python rag/scripts/ingest.py --dry-run
    python rag/scripts/ingest.py --verbose
"""

import sys
import json
import argparse
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent
sys.path.insert(0, str(PROJECT_ROOT))

from rag.src.scanner import scan_knowledge_files
from rag.src.normalizer import normalize_document
from rag.src.chunker import chunk_markdown


def main():
    parser = argparse.ArgumentParser(description="知识文件扫描预览")
    parser.add_argument("--dry-run", action="store_true", default=True, help="仅预览不写入")
    parser.add_argument("--verbose", action="store_true", help="显示详细信息")

    args = parser.parse_args()

    files = scan_knowledge_files()

    print(f"\n扫描到 {len(files)} 个知识文件:\n")

    by_type = {"rule": 0, "reference": 0, "project": 0}
    for f in files:
        by_type[f["source_type"]] = by_type.get(f["source_type"], 0) + 1

    for st, count in sorted(by_type.items()):
        print(f"  {st}: {count}")

    if args.verbose:
        print("\n文件列表:")
        for f in sorted(files, key=lambda x: x["rel_path"]):
            print(f"  [{f['source_type']}] {f['rel_path']}")

    # 预览分块统计
    print("\n分块预览 (前 5 个文档):")
    seen_docs = set()
    count = 0
    for f in files:
        doc = normalize_document(f["rel_path"], f["source_type"])
        if doc and doc["doc_id"] not in seen_docs:
            seen_docs.add(doc["doc_id"])
            chunks = chunk_markdown(doc)
            print(f"  {doc['doc_id']}: {len(chunks)} chunks")
            count += 1
            if count >= 5:
                break


if __name__ == "__main__":
    main()

File: rag/scripts/ingest.js
/**
 * ingest.js — 知识入库 CLI
 *
 * 用法：
 *   node scripts/ingest.js                  # 扫描并入库所有知识
 *   node scripts/ingest.js --dry-run        # 仅扫描，不写入
 *   node scripts/ingest.js --verbose        # 显示详细信息
 */

import { scanKnowledgeFiles, loadSourceConfig } from '../src/scanner.js';
import { normalizeDocument } from '../src/normalizer.js';
import { chunkMarkdown } from '../src/chunker.js';

const args = process.argv.slice(2);
const dryRun = args.includes('--dry-run');
const verbose = args.includes('--verbose');

async function main() {
  console.log('=== 知识入库 ===\n');

  // 扫描文件
  const files = scanKnowledgeFiles();
  console.log(`扫描到 ${files.length} 个知识文件\n`);

  if (dryRun) {
    for (const f of files) {
      console.log(`  [${f.source_type}] ${f.path}`);
    }
    console.log(`\n[DRY RUN] 共 ${files.length} 个文件，未写入`);
    return;
  }

  // 处理每个文件
  let docCount = 0;
  let chunkCount = 0;

  for (const file of files) {
    const doc = normalizeDocument(file.path, file.source_type);
    if (!doc) continue;

    docCount++;
    const chunks = chunkMarkdown(doc._raw_content);
    chunkCount += chunks.length;

    if (verbose) {
      console.log(`\n📄 ${doc.title} (${doc.source_type})`);
      console.log(`   路径: ${file.path}`);
      console.log(`   分类: ${doc.category}`);
      console.log(`   阶段: ${doc.stage.join(', ')}`);
      console.log(`   标签: ${doc.tags.slice(0, 5).join(', ')}`);
      console.log(`   Chunks: ${chunks.length}`);
      for (const c of chunks) {
        console.log(`     - [${c.chunk_type}] ${c.title} (${c.text.length}字)`);
      }
    } else {
      process.stdout.write('.');
      if (docCount % 20 === 0) console.log(` ${docCount}`);
    }
  }

  if (!verbose) console.log(` ${docCount}`);

  console.log(`\n入库完成!`);
  console.log(`  文档: ${docCount}`);
  console.log(`  Chunks: ${chunkCount}`);
  console.log(`\n运行 "node scripts/build-index.js" 来构建索引`);
}

main().catch(err => {
  console.error('入库失败:', err);
  process.exit(1);
});
